### PR TITLE
DM-8577: Add pybind11 rules to DM dev-guide

### PR DESCRIPTION
**Pull request by pschella**

### CODE REVIEW

**Comment by jonathansick** [comment]
Draft URL: https://developer.lsst.io/v/DM-8577/coding/pybind11_style_guide.html

File: coding/pybind11_style_guide.rst [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+The only classes that should be wrapped with ``unique_ptr`` are nonpolymorphic classes that are always passed by value or reference in C++ and are small enough that ``shared_ptr`` represents a significant overhead.
+
+Note that this does not mean that ``shared_ptr`` must be used in C++ code in preference to other options; the :ref:`C++ coding guidelines on when to use them <style-guide-cpp-5-24b>` still apply.
+
+.. _style-guide-pybind11-keyword-arguments:
+
+Keyword names MUST be provided for functions with multiple overloads or more than two arguments
+-----------------------------------------------------------------------------------------------
+
+Keywords arguments make Python code significantly more readable, especially when distinguishing between overloads or in long function signatures.
```
Comment (ktlim) [inline coding/pybind11_style_guide.rst]: "Keywords" -> "keyword"

File: coding/pybind11_style_guide.rst [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+
+Enums that require binary comparisons SHALL use py::arithmetic
+--------------------------------------------------------------
+
+If enums exposed to Python are used with bitwise comparison the desired operators shall not be added
+manually. Instead ``py::enum_<ExampleEnum>(..., py::arithmetic())`` SHALL be used.
+
+.. _style-guide-pybind11-internal-data-member-access:
+
+The ``reference_internal`` policy SHALL be used for functions (or properties) giving modifyable access to internal data members
```
Comment (ktlim) [inline coding/pybind11_style_guide.rst]: "modifyable" -> "modifiable" or even perhaps "write"

File: coding/pybind11_style_guide.rst [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+
+In rare cases, ``py::return_value_policy::reference_internal`` may be used if the expense of copying the object is large
+and the likelihood of accidental modification is low.
+
+.. _style-guide-pybind11-properties:
+
+All rules from the Python style guide regarding properties SHALL also apply to C++ wrappers
+-------------------------------------------------------------------------------------------
+
+See section (add reference) in the DM Python Style Guide.
```
Comment (ktlim) [inline coding/pybind11_style_guide.rst]: Maybe say instead "These rules are currently under development"?

File: coding/pybind11_style_guide.rst [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+        .export_values();
+
+* Scoped enums (i.e. ``enum class`` in C++) SHALL NOT use ``.export_values``.
+
+.. _style-guide-pybind11-enums-as-integers:
+
+Enums used as integers SHALL be wrapped as integer attributes
+-------------------------------------------------------------
+
+Regular (non-class) enums are frequently used in C++ to define a set of related integer constants rather than an actual enumeration.  These enums should be wrapped as simple integer class attributes rather than pybind11 enums, e.g.::
```
Comment (TallJimbo) [inline coding/pybind11_style_guide.rst]: I think it's worth explicitly mentioning bit flags here, as I think that's by far the most common case of enums used as integers, such as, "Enums whose values are defined to be distinct bits (e.g. `0x01, 0x02, 0x04`) are almost certainly used only as integer constants."

File: coding/pybind11_style_guide.rst [hunk 5]
```diff
@@ -0,0 +1,592 @@
...
+
+Using pure-Python modules only when necessary minimizes the number of source files and helps keep class definitions together.
+
+.. _style-guide-pybind11-include:
+
+Pybind11 headers should precede all other headers in the include ordering
+-------------------------------------------------------------------------
+
+``pybind11.h`` includes ``Python.h`` and must hence be included before all other headers.
+To keep a reasonable grouping, all other pybind11 headers should be included in this same include block.
```
Comment (TallJimbo) [inline coding/pybind11_style_guide.rst]: Should link to https://docs.python.org/3/c-api/intro.html#include-files for explanation.

**Pull request merged**